File: FFT/Lab3/Common/fortran/acchelper.f90

module ACCHELPER

    implicit none
    
    !Signal generation data
    real, parameter :: TWO_PI  = 2 * 3.14159265358979323846
    integer, parameter :: SIGNAL_SIZE = 67108864
    integer, parameter :: NUM_SAMPLES  = 44100
    integer, parameter :: FREQUENCY  = 100
    integer, parameter :: ITERS  = 100
    
    !store the frequencies loaded from the file
    integer, dimension(ITERS) :: h_freqArray
    
    ! Host/Device pointers used for cufft input and output
    double complex, allocatable, dimension(:) :: h_inputFFT, h_outputFFT
    
    !Timing variables
    real :: initTime, fftTime, diffTime, execTime, start, finish
    
    contains
        subroutine printMeanError(h_errors)
            implicit none

            real*8 :: errorSum = 0.0
            integer :: i
            double complex, allocatable, dimension(:), intent(in) :: h_errors
            
            !Adding all the differences and printing the mean error
            !This loop runs on the CPU
            do i=1, SIGNAL_SIZE
                errorSum = errorSum + real(h_errors(i))
            end do
            print *, "Mean Error: ", errorSum / SIGNAL_SIZE
        end subroutine printMeanError
end module ACCHELPER

File: FFT/Lab3/Common/c/acchelper.c
#include "acchelper.h"

size_t MEM_SIZE = sizeof(Complex) * SIGNAL_SIZE;

void allocate() {
 
    h_inputFFT = (Complex *)malloc(MEM_SIZE);
    h_outputFFT = (Complex *)malloc(MEM_SIZE);
}

void printMeanError(Complex *h_errors) {
    
    int i;
    double errorSum = 0.0;
    //Adding the differences
    for (i=0; i<SIGNAL_SIZE; ++i)
        errorSum += h_errors[i].x;
    //Show the Mean Error
    printf("\n\nMean Error =  %e", errorSum / SIGNAL_SIZE);
}

void deallocate() {
 
    free(h_inputFFT);
    free(h_outputFFT);
}